ibis/formats/polars.py

```diff
@@ -0,0 +1,176 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+import polars as pl
+
+import ibis.expr.datatypes as dt
+from ibis.expr.schema import Schema
+from ibis.formats import DataMapper, SchemaMapper, TableProxy, TypeMapper
+
+if TYPE_CHECKING:
+    from collections.abc import Sequence
+
+    import pandas as pd
+    import pyarrow as pa
+
+
+_to_polars_types = {
+    dt.Boolean: pl.Boolean,
+    dt.Null: pl.Null,
+    dt.String: pl.Utf8,
+    dt.Binary: pl.Binary,
+    dt.Date: pl.Date,
+    dt.Time: pl.Time,
+    dt.Int8: pl.Int8,
+    dt.Int16: pl.Int16,
+    dt.Int32: pl.Int32,
+    dt.Int64: pl.Int64,
+    dt.UInt8: pl.UInt8,
+    dt.UInt16: pl.UInt16,
+    dt.UInt32: pl.UInt32,
+    dt.UInt64: pl.UInt64,
+    dt.Float32: pl.Float32,
+    dt.Float64: pl.Float64,
+}
+
+_from_polars_types = {v: k for k, v in _to_polars_types.items()}
+
+
+class PolarsType(TypeMapper):
+    @classmethod
+    def to_ibis(cls, typ: pl.DataType, nullable=True) -> dt.DataType:
+        """Convert a polars type to an ibis type."""
+
+        base_type = typ.base_type()
+        if base_type is pl.Categorical:
+            return dt.String(nullable=nullable)
+        elif base_type is pl.Decimal:
+            return dt.Decimal(
+                precision=typ.precision, scale=typ.scale, nullable=nullable
+            )
+        elif base_type is pl.Datetime:
+            try:
+                timezone = typ.time_zone
+            except AttributeError:  # pragma: no cover
+                timezone = typ.tz  # pragma: no cover
+            return dt.Timestamp(timezone=timezone, nullable=nullable)
+        elif base_type is pl.Duration:
+            try:
+                time_unit = typ.time_unit
+            except AttributeError:  # pragma: no cover
+                time_unit = typ.tu  # pragma: no cover
+            return dt.Interval(unit=time_unit, nullable=nullable)
+        elif base_type is pl.List:
+            return dt.Array(cls.to_ibis(typ.inner), nullable=nullable)
+        elif base_type is pl.Struct:
+            return dt.Struct.from_tuples(
+                [(field.name, cls.to_ibis(field.dtype)) for field in typ.fields],
+                nullable=nullable,
+            )
+        else:
+            return _from_polars_types[base_type](nullable=nullable)
+
+    @classmethod
+    def from_ibis(cls, dtype: dt.DataType) -> pl.DataType:
+        """Convert an ibis type to a polars type."""
+        if dtype.is_decimal():
+            return pl.Decimal(
+                precision=dtype.precision,
+                scale=9 if dtype.scale is None else dtype.scale,
+            )
+        elif dtype.is_timestamp():
+            return pl.Datetime("ns", dtype.timezone)
```

Comment:
Since the ibis timestamps have a scale we can do more accurate conversions here using `dtype.unit.short`. I assume this was the previous conversion logic from the backend implementation. Since it requires additional testing it could you please create a follow-up ticket so we improve this later?